### PR TITLE
[UNDERTOW-2028] Buffer leak error involving AjpServerResponseConduit

### PR DESCRIPTION
**Pull request by rmartinc**
Issue: https://issues.redhat.com/browse/UNDERTOW-2028

Changing `AjpServerResponseConduit` to manage responses bigger than the buffer size. Now if the conduit cannot write more to the network it stores several buffers to accommodate the remaining data. If any weird error happens the buffers are closed properly.

### CODE REVIEW

**Review by fl4via — APPROVED** [review]

File: core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java [hunk 1]
```diff
@@ -266,6 +268,40 @@ protected void queueCloseFrames() {
         queueFrame(null, buffer);
     }
 
+    private void queueRemainingBytes(final ByteBuffer src, final ByteBuffer[] buffers) {
+        List<PooledByteBuffer> pools = new ArrayList<>(4);
+
+        try {
+            PooledByteBuffer newPooledBuffer = pool.allocate();
```
Comment (fl4via) [inline core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java:275]: This buffer is leaking as can be seen here: https://github.com/undertow-io/undertow/runs/5754811783?check_suite_focus=true (windows failure)

File: core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java [hunk 2]
```diff
@@ -266,6 +268,40 @@ protected void queueCloseFrames() {
...
+
+            ByteBuffer[] savedBuffers = new ByteBuffer[pools.size() + 2];
+            int i = 0;
+            savedBuffers[i++] = buffers[0];
+            for (PooledByteBuffer p : pools) {
+                savedBuffers[i++] = p.getBuffer();
+            }
+            savedBuffers[i] = buffers[2];
+            queueFrame(new PooledBuffersFrameCallback(pools.toArray(new PooledByteBuffer[0])), savedBuffers);
+        } catch (RuntimeException | Error e) {
```
Comment (fl4via) [inline core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java:297]: better use a finally block

**Review by fl4via — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java (hunk 2 above)
Comment (rmartinc) [reply core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java:297]: Hi @fl4via! 

I think the problem is not at this point. It's what I tried to explain yesterday in the meeting. The buffers here are saved because writes returned 0 (no more writing) and they need to be queued to send them later (when writes are called again). So a finally block to close them is not valid, they should be created now and closed when fully sent. My feeling is that the request is somehow hanging and still on-going in the backend server but the proxy timed out (note the test takes more than 300s, and it usually is around 20s). Other thing is why the test is a success, which is very strange. But it can be a ConnectionClosedException which is ignored [here](https://github.com/undertow-io/undertow/blob/master/servlet/src/test/java/io/undertow/servlet/test/streams/ServletOutputStreamTestCase.java#L271-L275). But I'm not sure about this part.

In summary, I think we have two separate issues involving AjpServerResponseConduit, and this PR just fixes the easy one. We need to investigate what happens later and why the buffer is not closed when sent. It can be because it hangs (maybe related to UNDERTOW-1945 or UNDERTOW-1948) or a missing condition on close (but I have not seen any obvious mistake yet).

What do you think about creating a followup JIRA and at least merging this fix?

Comment (fl4via) [reply core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java:297]: @rmartinc thanks for the explanation. That sounds fine.

Comment (fl4via) [reply core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java:297]: @rmartinc  I'm not sure if the failure I saw a couple of days ago was the same. I have just created https://issues.redhat.com/browse/UNDERTOW-2054 to track the Ssl buffer leak errors on Windows proxy-https tests. But I have the impression that, when we were discussing this, we were talking about a buffer created by AjpServerResponseConduit, can yo confirm? As soon as I see that failure again, I'll create the Jira.

Comment (fl4via) [reply core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java:297]: You can disregard this comment, I found the issue again and I saw your comments in the Jira, here: https://issues.redhat.com/browse/UNDERTOW-2028?focusedCommentId=20027058&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-20027058
Let's proceed with UNDERTOW-1945 as it seems it is the same failure indeed. Thanks for having a look at this!
